Add Challenges 20 21 22 for str Pattern and iter (#266)
model-checking/verify-rust-std: doc/src/challenges/0020-str-pattern-pt1.md

- **End date:** *2025-10-17*
- **Reward:** *25000 USD*

-------------------
## Goal
Verify the safety of char-related `Searcher` methods in `str::pattern`.

## Motivation

String and `str` types are widely used in Rust programs, so it is important that their associated functions do not cause undefined behavior.

## Description

The following str library functions are generic over the `Pattern` trait (https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html): 
- `contains`
- `starts_with`
- `ends_with`
- `find` 
- `rfind`
- `split`
- `split_inclusive`
- `rsplit`
- `split_terminator`
- `rsplit_terminator`
- `splitn`
- `rsplitn`
- `split_once`
- `rsplit_once`
- `rmatches`
- `match_indices`
- `rmatch_indices`
- `trim_matches`
- `trim_start_matches`
- `strip_prefix`
- `strip_suffix`
- `trim_end_matches`
These functions accept a pattern as input, then call [into_searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#tymethod.into_searcher) to create a [Searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#associatedtype.Searcher) for the pattern. They use this `Searcher` to perform their desired operations (split, find, etc.).

> Those functions are implemented in (library/core/src/str/mod.rs), but the core of them is the searching algorithms which are implemented in (library/core/src/str/pattern.rs).

Reviewer: ```suggestion
Those functions are implemented in (library/core/src/str/mod.rs), but the core of them are the searching algorithms which are implemented in (library/core/src/str/pattern.rs).
```

# Challenge 20: Verify the safety of char-related functions in str::pattern

- **Status:** Open
- **Tracking Issue:** [#29](https://github.com/model-checking/verify-rust-std/issues/29)
- **Start date:** *2025/03/07*

> - **End date:** *2025/10/17*

Reviewer: See my comments on https://github.com/model-checking/verify-rust-std/pull/260.

> - **Reward:** *5,000 USD*

Reviewer: Can you please confirm that we have committee approval for this?

Author: The dates, tracking link and reward will be fixed after the challenges are approved. Please focus on the content first.

-------------------

### Context

A majority portion in str library functions take a `Pattern` (https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html) as input. 
The functions which take Pattern as input turn the input str into a kind of `Searcher` (https://doc.rust-lang.org/std/str/pattern/trait.Searcher.html)  which iterates over positions where the Pattern match, then perform their desired operations (split, find, ...).

> Those functions are implemented in (library/core/src/str/mod.rs), but the core of them is the searching algorithms which are implemented in (library/core/src/str/pattern.rs).

Reviewer 2: I don't understand this section; I don't see any functions in `mod.rs` that take in a string and turn that string into a `Searcher` associated type. There's syntax like this:
https://github.com/model-checking/verify-rust-std/blob/9b27cfbb4ff8f21230c6c7afa899f1c71eccaba2/library/core/src/str/mod.rs#L1391-L1393

but that specifies a [higher-order trait bound](https://doc.rust-lang.org/stable/reference/trait-bounds.html#higher-ranked-trait-bounds) over the generic type `P` (and these functions don't take a string as input).

Author: Yes, it is not obvious in the code of the function, but sooner or later, the function into_searcher(self, haystack: &str) -> Self::Searcher<'_> is called to create a Searcher. In your example, it is called in the function is_suffix_of.

# Details

**Important note:** for this challenge, you can assume: 
1. The safety and functional correctness of all functions in `slice` module 
2. That all functions in (library/core/src/str/validations.rs) are functionally correct (consistent with the UTF8 encoding description in https://en.wikipedia.org/wiki/UTF-8). 
3. That all the Searchers in (library/core/src/str/iter.rs) are created by the into_searcher(_, haystack) with haystack being a valid utf8 string (str). You can assume any utf8 string property of haystack.

Verify the safety of the functions in (library/core/src/str/pattern.rs) listed in the next section.

The safety properties we are targeting are: 

> 1. No UB occurs when calling the functions after the Searcher is created.

Reviewer 2: ```suggestion
1. No undefined behavior occurs when calling the functions after the Searcher is created.
```

Reviewer 2: Also, which functions?

2. The impls of unsafe traits `Searcher` and `ReverseSearcher` satisfies the SAFETY condition stated in the file: 
```
/// The trait is marked unsafe because the indices returned by the
/// [`next()`][Searcher::next] methods are required to lie on valid utf8
/// boundaries in the haystack. This enables consumers of this trait to
/// slice the haystack without additional runtime checks.
```
This property should hold for next_back() of `ReverseSearcher` too.

### Success Criteria

Verify the safety of the following functions in (library/core/src/str/pattern.rs) : `next`, `next_match`, `next_back`, `next_match_back`, `next_reject`, `next_back_reject`
which are implemented for the following `Searcher`s:  `CharSearcher`, `MultiCharEqSearcher`, `CharArraySearcher` , `CharArrayRefSearcher`, `CharSliceSearcher`, `CharPredicateSearcher`.

> The verification is consider successful if for each `Searcher` above, you can specify a condition (a "type invariant") C and prove that:

Reviewer: I'd prefer not to use quotes as each reader may interpret those differently. If it is the right technical term (I think so), preferably also include a pointer to a reference where people can learn more about it.

> The verification is consider successful if for each `Searcher` above, you can specify a condition (a "type invariant") C and prove that:

Reviewer 2: ```suggestion
The verification is considered successful if for each `Searcher` above, you can specify a condition (a "type invariant") C and prove that:
```

1. If the `Searcher` is created from any valid utf8 haystack, it satisfies C.
2. If the `Searcher` satisfies C, it ensures the two safety properties mentioned in the previous section.
3. If the `Searcher` satisfies C, after it calls any function above and gets modified, it still statisfies C.

The verification must be unbounded---it must hold for inputs of arbitrary size.

### List of UBs

All proofs must automatically ensure the absence of the following undefined behaviors [ref](https://github.com/rust-lang/reference/blob/142b2ed77d33f37a9973772bd95e6144ed9dce43/src/behavior-considered-undefined.md):

* Accessing (loading from or storing to) a place that is dangling or based on a misaligned pointer.
* Reading from uninitialized memory except for padding or unions.
* Mutating immutable bytes.

> * Producing an invalid value

Reviewer 2: Do these UBs make the most sense given this challenge? You may want to pick different ones from the list to customize it to the challenge.

# Challenge 20: Verify the safety of char-related functions in str::pattern

- **Status:** Open
- **Tracking Issue:** [#277](https://github.com/model-checking/verify-rust-std/issues/277)
- **Start date:** *2025-03-07*
- **End date:** *2025-10-17*

> - **Reward:** *5000 USD*

Reviewer 2: This strikes me as a more difficult challenge--should we say $10,000? What do you think?

Author: I agree.

strip_prefix, strip_suffix, trim_end_matches.
The functions which take Pattern as input turn the input str into a kind of `Searcher` (https://doc.rust-lang.org/std/str/pattern/trait.Searcher.html)  which iterates over positions where the Pattern match, then perform their desired operations (split, find, ...).
Those functions are implemented in (library/core/src/str/mod.rs), but the core of them is the searching algorithms which are implemented in (library/core/src/str/pattern.rs).

# Details

**Important note:** for this challenge, you can assume: 
1. The safety and functional correctness of all functions in `slice` module.
2. That all functions in (library/core/src/str/validations.rs) are functionally correct (consistent with the UTF8 encoding description in https://en.wikipedia.org/wiki/UTF-8). 
3. That all the Searchers in (library/core/src/str/iter.rs) are created by the into_searcher(_, haystack) with haystack being a valid utf8 string (str). You can assume any utf8 string property of haystack.

Verify the safety of the functions in (library/core/src/str/pattern.rs) listed in the next section.

The safety properties we are targeting are: 

> 1. No undefined behavior occurs when calling the functions after the Searcher is created.

Reviewer 2: ```suggestion
1. No undefined behavior occurs after the Searcher is created.
```

2. The impls of unsafe traits `Searcher` and `ReverseSearcher` satisfy the SAFETY condition stated in the file: 
```
/// The trait is marked unsafe because the indices returned by the
/// [`next()`][Searcher::next] methods are required to lie on valid utf8
/// boundaries in the haystack. This enables consumers of this trait to
/// slice the haystack without additional runtime checks.
```
This property should hold for next_back() of `ReverseSearcher` too.

### Success Criteria

> Verify the safety of the following functions in (library/core/src/str/pattern.rs) : next, next_match, next_back, next_match_back, next_reject, next_back_reject
> which are implemented for the following `Searcher`s:  CharSearcher, MultiCharEqSearcher, CharArraySearcher , CharArrayRefSearcher, CharSliceSearcher, CharPredicateSearcher.

Reviewer 2: ```suggestion
Verify the safety of the following functions in (library/core/src/str/pattern.rs) : 
- `next`
- `next_match`
- `next_back`
- `next_match_back`
- `next_reject`
- `next_back_reject`
for the following `Searcher`s: 
- `CharSearcher`
- `MultiCharEqSearcher`
- `CharArraySearcher`
- `CharArrayRefSearcher`
- `CharSliceSearcher`
- `CharPredicateSearcher`
```

I deleted "which are implemented for" because that implies its a complete list, but it's not because we save the `StrSearcher` implementation for the next challenge.

> The verification is considered successful if for each `Searcher` above, you can specify a condition (a "type invariant") C and prove that:
> 1. If the `Searcher` is created from any valid utf8 haystack, it satisfies C.
> 2. If the `Searcher` satisfies C, it ensures the two safety properties mentioned in the previous section.
> 3. If the `Searcher` satisfies C, after it calls any function above and gets modified, it still statisfies C.

Reviewer 2: ```suggestion
The verification is considered successful if for each `Searcher` above, you can specify a condition (a "type invariant") `C` and prove that:
1. If the `Searcher` is created from any valid UTF-8 haystack, it satisfies `C`.
2. If the `Searcher` satisfies `C`, it ensures the two safety properties mentioned in the previous section.
3. If the `Searcher` satisfies `C`, after it calls any function above and gets modified, it still satisfies `C`.
```

- **Tracking Issue:** [#277](https://github.com/model-checking/verify-rust-std/issues/277)
- **Start date:** *2025-03-07*
- **End date:** *2025-10-17*
- **Reward:** *10000 USD*

-------------------

> 

Reviewer 2: ```suggestion
## Goal
Verify the safety of `Searcher` methods in `str::pattern`.

```

> ### Context

Reviewer 2: Can you use the [challenge template](https://github.com/model-checking/verify-rust-std/blob/main/doc/src/challenge_template.md) headers instead of your own? We should be consistent across challenges.

(Also, this context header has three # and the details one has one #, so it looks odd when you render it).

Reviewer 2: As part of that, please add "motivation" sections to this challenge and the next ones. I think you can just condense "context" and "details" into one details section, but I don't mind if you'd prefer to keep them separate.

- **Reward:** *10000 USD*

-------------------
## Goal
Verify the safety of char-related `Searcher` methods in `str::pattern`.

## Motivation

> String and str types are widely used in Rust programs. Verifying Rust String and str functions in Rust standard library is important in ensuring the safety of these programs.

Reviewer 2: ```suggestion
String and `str` types are widely used in Rust programs, so it is important that their associated functions do not cause undefined behavior.
```

These functions accept a pattern as input, then call [into_searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#tymethod.into_searcher) to create a [Searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#associatedtype.Searcher) for the pattern. They use this `Searcher` to perform their desired operations (split, find, etc.).
Those functions are implemented in (library/core/src/str/mod.rs), but the core of them is the searching algorithms which are implemented in (library/core/src/str/pattern.rs).

> ### Assumption

Reviewer 2: ```suggestion
### Assumptions
```
